addons/account/models/account_bank_statement.py

```diff
@@ -83,6 +83,28 @@ def _is_difference_zero(self):
         for bank_stmt in self:
             bank_stmt.is_difference_zero = float_is_zero(bank_stmt.difference, precision_digits=bank_stmt.currency_id.decimal_places)
 
+    @api.depends('date', 'balance_start', 'balance_end_real')
+    def _compute_is_valid_balance_start(self):
+        domain = [('journal_id.type', '=', 'bank')]
+        for bank_stmt in self:
+            record_id = self.env.context.get('record_id')
+            # Wierd behaviour, if I edit the already saved record and modify any of it's fields from depends then it converts `bank_stmt` into NewId() recordset
+            # without entering into edit mode, it gives regular recordset, that is why id is passed from view side in field context
+            if record_id or isinstance(bank_stmt.id, pycompat.integer_types):
+                previous_stmt = self.search(domain + [('date', '=', bank_stmt.date), ('id', '<', record_id or bank_stmt.id)], limit=1)
+                if not previous_stmt:
+                    previous_stmt = self.search(domain + [('date', '<', bank_stmt.date), ('id', '!=', record_id or bank_stmt.id)], limit=1)
+            else:
+                # Since you are here it simply means you have NewID() recordset in `bank_stmt` and you are in creation mode
+                previous_stmt = self.search(domain + [('date', '<=', bank_stmt.date)], limit=1)
```

Comment:
I think we should not use `create_date` field for our domain to get the previous statement because an user can create bank statements with any previous dates at any time and this may result in giving wrong previous statement. For example, in below scenario it would fail to give correct result,

In database without demo data(not sure though how much below case is practical),
create first statement with statement date as 3rd Mar, 2018(starting balance = 400, ending balance = 600(encode one statement line with 300$))
create second statement with statement date as 1st Mar, 2018(starting balance = 0, ending balance = 100(encode one statement line with 100$))
create third statement with statement date as 2nd Mar, 2018(starting balance = 100, ending balance = 300(encode one statement line with 200$))

In the above case, it should show warning message on statement with date 3rd March since it should have starting balance 300(instead of 400), but it will not show warning.

The reason why I passed the id that way is because when we have several statements with same statement date, at that time the statement's ID comes into the picture as the default order of model `account.bank.statement` is `date desc, id desc`. So, if we have statements like,

ID, statement date, starting balance, ending balance
13, 03-05-2018, 6000, 9000
12, 03-05-2018, 4000, 6000
11, 03-05-2018, 2000, 4000
10, 03-05-2018, 1000, 2000


and, if I need to find which one is the previous statement for statement with ID 12, I have to check the ID of statement along with date, like https://github.com/odoo/odoo/pull/18664/files#diff-86c8da67df78eedb5113360f754e1f75R94. Do you think there can be other ways to do it ?

Reply:
You're confusing date and create_date fields ;) create_date is inherited by every model and cannot be set by the user, it is automatically filled by the ORM when creating the record. Furthermore, it is a datetime, so it is practically impossible to have two statements with the exact same create_date.

Reply:
@oco-odoo I am still not sure that using `create_date` in our domain would help us get our desire result. For now, I added a new compute field that is having the code you suggested and also I kept the code that I did so that we can have the comparison side by side on runbot. <a href="https://github.com/odoo/odoo/pull/18664/commits/de77a43d948f75eda3b7002d48ee45f4f7c0b128">Here's</a> the implementation of your suggestion and here's the <a href="http://309644-master-account-bank-statement-da-de77a4.runbot8.odoo.com/web/?debug#view_type=list&model=account.bank.statement&action=148&active_id=7">runbot case</a> where it is failing to give desire result on runbot.
Can you please have a look at it ?
Thanks

Reply:
You're right, I had misunderstood your previous remark. Indeed using create_date will not work here. What we simply can do however, is just use date, and rely on the _order defined on bank statements to return the most recent statement (i.e. the one with the greatest id) in case multiple statements were made on the same date. I just pushed a new version of the "oco-way" method doing that. What do you say? :)

Reply:
@oco-odoo adding `('id', '!=', bank_stmt.id)` in domain will again give us wrong result when there are multiple statements created on same date.
Assume we have 2 statements created on 12 March,
```
ID, name, date, start balance, end balance
5, 12Mar-1, 03-12-2018, 200, 400
4, 12Mar-2, 03-12-2018, 100, 200
3, 03Mar, 03-03-2018, 50, 100
```
and we are on statement with ID `#4` then search with domain
`[('journal_id.type', '=', 'bank'), ('date', '<=', '2018-03-12'), ('journal_id', '=', XX), ('id', '!=', 4)]` will result in giving statement with ID `#5` as previous statement because of `date desc, id desc` default model ordering. Actually this was the reason why I had to use the `<` operator with `ID` field and since framework gives NewID for compute function when editing the record, I used the trick to pass it from XML that way.
Here I implemented that case on <a href="http://309865-18664-ab31cc.runbot9.odoo.com/web#view_type=list&model=account.bank.statement&action=148&active_id=7">runbot branch</a> after the changes you applied. Can you please have a look at it again.

Reply:
Indeed... Don't you think it would then be easier to simply use a datetime instead of a date for the field?

Reply:
do you mean to change the type of `date`(ie. Bank Statement's date) from `date` to `datetime` ?

Reply:
Precisely. It would totally make sense in this context, after all.